x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutJobAction.java

```diff
@@ -59,6 +59,10 @@ public static Request parseRequest(String jobId, XContentParser parser) {
         public Request(Job.Builder jobBuilder) {
             // Validate the jobBuilder immediately so that errors can be detected prior to transportation.
             jobBuilder.validateInputFields();
+            // Validate that detector configs are unique.
+            // This validation logically belongs to validateInputFields call but we perform it only for PUT action to avoid BWC issues which
+            // would occur when parsing an old job config that already had duplicate detectors.
+            jobBuilder.validateDetectorsAreUnique();
```

Comment:
@droberts195 Since we now store the jobs out of the cluster state, could we move this validation in `build()`? Also, should we target this PR only against `8.0` as there still might be jobs in cluster state in `7.x` nodes?

Reply:
I don't think we can ever move this validation inside `Job.Builder.build()`.  Once our configs are out of cluster state we won't break the whole cluster if we put the validation there, but we _will_ still break ML if people have jobs that don't conform to the new validation rules.  So the best we can do is to add validation that newly added jobs conform to the tighter rules.  This PR is basically implementing the idea of https://github.com/elastic/elasticsearch/issues/39704#issuecomment-470091066.